In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set path

path = r'C:\Users\Josh Wattay\OneDrive\Documents\Desktop\Flight Delay Analysis - Raw Data\2023'

In [3]:
# Import Datasets

df = pd.read_csv(os.path.join(path, 'flights2023.csv'), index_col = False)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_15620\3299958900.py:3: DtypeWarning: Columns (11,52) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(path, 'flights2023.csv'), index_col = False)


In [4]:
# Shape Check
df.shape

(7278739, 70)

In [5]:
# Set the display option to show all 70 columns
pd.set_option('display.max_columns', 70)

In [6]:
#Inspect Dataframe and begin Cleaning by ensuring columns are named appropriately
df.head()

Unnamed: 0  Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0           0  2023        1      1          22          7  2023-01-22   
1           1  2023        1      1          22          7  2023-01-22   
2           2  2023        1      1          22          7  2023-01-22   
3           3  2023        1      1          22          7  2023-01-22   
4           4  2023        1      1          22          7  2023-01-22   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                             1447   
1                          B6                             1451   
2                          B6                             1453   
3                          B6                             1454   
4                          B6                             1455   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 B6   
1                                               NaN                 B6   
2                                               NaN                 B6   
3                                               NaN                 B6   
4                                               NaN                 B6   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20409                          B6      N636JB   
1                     20409                          B6      N3112J   
2                     20409                          B6      N507JT   
3                     20409                          B6      N956JT   
4                     20409                          B6      N249JB   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             1447            12197             1219702   
1                             1451            10721             1072102   
2                             1453            11697             1169706   
3                             1454            14843             1484306   
4                             1455            10721             1072102   

   OriginCityMarketID Origin       OriginCityName OriginState  \
0               31703    HPN     White Plains, NY          NY   
1               30721    BOS           Boston, MA          MA   
2               32467    FLL  Fort Lauderdale, FL          FL   
3               34819    SJU         San Juan, PR          PR   
4               30721    BOS           Boston, MA          MA   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0               36        New York          15304           1530402   
1               25   Massachusetts          13204           1320402   
2               12         Florida          14843           1484306   
3               72     Puerto Rico          11697           1169706   
4               25   Massachusetts          11278           1127805   

   DestCityMarketID Dest         DestCityName DestState  DestStateFips  \
0             33195  TPA            Tampa, FL        FL             12   
1             31454  MCO          Orlando, FL        FL             12   
2             34819  SJU         San Juan, PR        PR             72   
3             32467  FLL  Fort Lauderdale, FL        FL             12   
4             30852  DCA       Washington, DC        VA             51   

  DestStateName  CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0       Florida         800    756.0      -4.0              0.0       0.0   
1       Florida        1955   2048.0      53.0             53.0       1.0   
2

The 'Unnamed: 0' column will be removed as it is an additional index and redundant. 

In [7]:
# Drop the 'Unnamed: 0' column
df = df.drop(columns=['Unnamed: 0'])

In [8]:
df.shape

(7278739, 69)

In [9]:
df.head()

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0  2023        1      1          22          7  2023-01-22   
1  2023        1      1          22          7  2023-01-22   
2  2023        1      1          22          7  2023-01-22   
3  2023        1      1          22          7  2023-01-22   
4  2023        1      1          22          7  2023-01-22   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                             1447   
1                          B6                             1451   
2                          B6                             1453   
3                          B6                             1454   
4                          B6                             1455   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 B6   
1                                               NaN                 B6   
2                                               NaN                 B6   
3                                               NaN                 B6   
4                                               NaN                 B6   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20409                          B6      N636JB   
1                     20409                          B6      N3112J   
2                     20409                          B6      N507JT   
3                     20409                          B6      N956JT   
4                     20409                          B6      N249JB   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             1447            12197             1219702   
1                             1451            10721             1072102   
2                             1453            11697             1169706   
3                             1454            14843             1484306   
4                             1455            10721             1072102   

   OriginCityMarketID Origin       OriginCityName OriginState  \
0               31703    HPN     White Plains, NY          NY   
1               30721    BOS           Boston, MA          MA   
2               32467    FLL  Fort Lauderdale, FL          FL   
3               34819    SJU         San Juan, PR          PR   
4               30721    BOS           Boston, MA          MA   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0               36        New York          15304           1530402   
1               25   Massachusetts          13204           1320402   
2               12         Florida          14843           1484306   
3               72     Puerto Rico          11697           1169706   
4               25   Massachusetts          11278           1127805   

   DestCityMarketID Dest         DestCityName DestState  DestStateFips  \
0             33195  TPA            Tampa, FL        FL             12   
1             31454  MCO          Orlando, FL        FL             12   
2             34819  SJU         San Juan, PR        PR             72   
3             32467  FLL  Fort Lauderdale, FL        FL             12   
4             30852  DCA       Washington, DC        VA             51   

  DestStateName  CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0       Florida         800    756.0      -4.0              0.0       0.0   
1       Florida        1955   2048.0      53.0             53.0       1.0   
2   Puerto Rico        1727   1734.0       7.0              7.0       0.0

Rename syntax for various columns for consistency and accuracy.

In [10]:
# Define the dictionary for renaming columns
rename_dict = {
    'DayofMonth': 'Day_of_Month',
    'DayOfWeek': 'Day_of_Week',
    'FlightDate': 'Flight_Date',
    'Operating_Airline': 'Operating Airline',
    'OriginAirportID': 'Origin_Airport_ID',
    'OriginAirportSeqID': 'Origin_Airport_Seq_ID',
    'OriginCityMarketID': 'Origin_City_Market_ID',
    'OriginCityName': 'Origin_City_Name',
    'OriginState': 'Origin_State',
    'OriginStateFips': 'Origin_State_Fips',
    'OriginStateName': 'Origin_State_Name',
    'DestAirportID': 'Dest_Airport_ID',
    'DestAirportSeqID': 'Dest_Airport_Seq_ID',
    'DestCityMarketID': 'Dest_City_Market_ID',
    'DestCityName': 'Dest_City_Name',
    'DestState': 'Dest_State',
    'DestStateFips': 'Dest_State_Fips',
    'DestStateName': 'Dest_State_Name',
    'CRSDepTime': 'CRS_Dep_Time',
    'DepTime': 'Dep_Time',
    'DepDelay': 'Dep_Delay',
    'DepDelayMinutes': 'Dep_Delay_Minutes',
    'DepDel15': 'Dep_Del_15',
    'DepartureDelayGroups': 'Departure_Delay_Groups',
    'DepTimeBlk': 'Dep_Time_Blk',
    'TaxiOut': 'Taxi_Out',
    'WheelsOff': 'Wheels_Off',
    'WheelsOn': 'Wheels_On',
    'TaxiIn': 'Taxi_In',
    'CRSArrTime': 'CRS_Arr_Time',
    'ArrTime': 'Arr_Time',
    'ArrDelay': 'Arr_Delay',
    'ArrDelayMinutes': 'Arr_Delay_Minutes',
    'ArrDel15': 'Arr_Del_15',
    'ArrivalDelayGroups': 'Arrival_Delay_Groups',
    'ArrTimeBlk': 'Arr_Time_Blk',
    'CancellationCode': 'Cancellation_Code',
    'CRSElapsedTime': 'CRS_Elapsed_Time',
    'ActualElapsedTime': 'Actual_Elapsed_Time',
    'AirTime': 'Air_Time',
    'DistanceGroup': 'Distance_Group',
    'CarrierDelay': 'Carrier_Delay',
    'WeatherDelay': 'Weather_Delay',
    'NASDelay': 'NAS_Delay',
    'SecurityDelay': 'Security_Delay',
    'LateAircraftDelay': 'Late_Aircraft_Delay',
    'FirstDepTime': 'First_Dep_Time',
    'TotalAddGTime': 'Total_Add_G_Time',
    'LongestAddGTime': 'Longest_Add_G_Time',
    'DivDistance': 'Div_Distance'
}

# Rename the columns
df = df.rename(columns=rename_dict)

In [11]:
df.shape

(7278739, 69)

In [12]:
# Drop the additional irrelevant columns - please refer to the 'Cleaning 2019 Flight Data' script for more details.

df = df.drop(columns=['First_Dep_Time', 'Total_Add_G_Time', 'Longest_Add_G_Time', 'Div_Distance'])

Fortunately, this raw data comes with a Duplicate flag column.
I will check the counts for this column to check for any values of 'Y' or Yes.
To be certain, I will also run a duplicate check with the .duplicated command.

In [13]:
df['Duplicate'].value_counts()

Duplicate
N    7278739
Name: count, dtype: int64

In [14]:
# So far so good, now to check for duplicates in the entire dataframe by creating a subset of duplicated rows

# Create a DataFrame containing only the duplicate rows
dups = df[df.duplicated(keep=False)]

# Display the 'dups' DataFrame
print(dups)

Empty DataFrame
Columns: [Year, Quarter, Month, Day_of_Month, Day_of_Week, Flight_Date, Marketing_Airline_Network, DOT_ID_Marketing_Airline, IATA_Code_Marketing_Airline, Flight_Number_Marketing_Airline, IATA_Code_Originally_Scheduled_Code_Share_Airline, Operating_Airline , DOT_ID_Operating_Airline, IATA_Code_Operating_Airline, Tail_Number, Flight_Number_Operating_Airline, Origin_Airport_ID, Origin_Airport_Seq_ID, Origin_City_Market_ID, Origin, Origin_City_Name, Origin_State, Origin_State_Fips, Origin_State_Name, Dest_Airport_ID, Dest_Airport_Seq_ID, Dest_City_Market_ID, Dest, Dest_City_Name, Dest_State, Dest_State_Fips, Dest_State_Name, CRS_Dep_Time, Dep_Time, Dep_Delay, Dep_Delay_Minutes, Dep_Del_15, Departure_Delay_Groups, Dep_Time_Blk, Taxi_Out, Wheels_Off, Wheels_On, Taxi_In, CRS_Arr_Time, Arr_Time, Arr_Delay, Arr_Delay_Minutes, Arr_Del_15, Arrival_Delay_Groups, Arr_Time_Blk, Cancelled, Cancellation_Code, Diverted, CRS_Elapsed_Time, Actual_Elapsed_Time, Air_Time, Flights, Distance,

In [15]:
# No dups!

Now I will address missing and/or null values.

In [16]:
# Set the display option to show all rows
pd.set_option('display.max_rows', None)

In [17]:
# Find the count of null values in each column
null_counts = df.isnull().sum()

In [18]:
# Display the count of null values
print(null_counts)

Year                                                       0
Quarter                                                    0
Month                                                      0
Day_of_Month                                               0
Day_of_Week                                                0
Flight_Date                                                0
Marketing_Airline_Network                                  0
DOT_ID_Marketing_Airline                                   0
IATA_Code_Marketing_Airline                                0
Flight_Number_Marketing_Airline                            0
IATA_Code_Originally_Scheduled_Code_Share_Airline    7277653
Operating_Airline                                          0
DOT_ID_Operating_Airline                                   0
IATA_Code_Operating_Airline                                0
Tail_Number                                            19648
Flight_Number_Operating_Airline                            0
Origin_Airport_ID       

In [19]:
# The IATA_Code_Originally_Scheduled_Code_Share column is nearly completely null
# Given that the Bureau of Transportation Statistics (BTS.gov) stated that this column shouldn't be 
# used for analysis, and it is completely null for 2019, I will be dropping it and keeping the Unique_Code
# column as suggested by BTS.gov
# Here is the language from BTS.gov:
# ""Code assigned by IATA and commonly used to identify a carrier.""
#""As the same code may have been assigned to different carriers over time, the code is not always unique.""
#""For analysis, use the Unique Carrier Code.""

# Drop the 'IATA_Code_Originally_Scheduled_Code_Share_Airline' column
df = df.drop(columns=['IATA_Code_Originally_Scheduled_Code_Share_Airline'])

In [20]:
df.shape

(7278739, 64)

There are a significant amount of null values for Dep_Delay, Taxi_Out, Taxi_In, Arr_Time.
I believe these are null values because the flights were cancelled.
I will check value counts for the Cancelled column. Value counts for '1' should be close if not the same to the values missing.

In [21]:
df['Cancelled'].value_counts()

Cancelled
0.0    7184842
1.0      93897
Name: count, dtype: int64

93897 flights were cancelled in 2023. 

This value is fairly close to the missing values in:

Dep_Time                                               90100
Dep_Delay                                              90160
Dep_Delay_Minutes                                      90160
Dep_Del_15                                             90160
Departure_Delay_Groups                                 9
Taxi_Out                                               93119
Wheels_Off                                             93119
Wheels_On                                              95647
Taxi_In                                                95
Arr_Time                                               95646
Arr_Delay                                             111695
Arr_Delay_Minutes                                     111695
Arr_Del_15                                            111695
Arrival_Delay_Groups                                  11

The missing values in:

Cancellation_Code                                    7184842 -- are accounted for. Total Flights (7278739) - Total Cancelled Flights (93897) = 7184842. 

After careful consideration, I believe it is best to create a cancelled flights flag in order to prevent confusion in future analysis for delays v. cancellation. 16956470160

In [22]:
# Identify rows where flights were cancelled

cancelled_flights = df['Cancelled'] == 1

In [23]:
# Create a column to indicate if the data is missing due to cancellation

df['Missing_Due_To_Cancellation'] = cancelled_flights

Now I will address columns containing mixed datatypes.

In [24]:
df.dtypes

Year                                 int64
Quarter                              int64
Month                                int64
Day_of_Month                         int64
Day_of_Week                          int64
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int64
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int64
Operating_Airline                   object
DOT_ID_Operating_Airline             int64
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int64
Origin_Airport_ID                    int64
Origin_Airport_Seq_ID                int64
Origin_City_Market_ID                int64
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                    int64
Origin_State_Name                   object
Dest_Airpor

In [25]:
# I will change many of these datatypes to reduce the amount of memory needed by reducing int64 to int8, 16, or 32 where possible.
# Additionally, I will change the Flight_Date column to a date datatype, instead of an object.
# Lets change Flight_Date first

# Ensure 'Flight_Date' column is in datetime format
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'], errors='coerce')

In [26]:
# Extract only the date part (since we don't have time in the same column)
df['Flight_Date'] = df['Flight_Date'].dt.date

In [27]:
# Verify the result
print(df.head())
print(df.dtypes)

   Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2023        1      1            22            7  2023-01-22   
1  2023        1      1            22            7  2023-01-22   
2  2023        1      1            22            7  2023-01-22   
3  2023        1      1            22            7  2023-01-22   
4  2023        1      1            22            7  2023-01-22   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                             1447   
1                          B6                             1451   
2                          B6                             1453

In [28]:
# Change datatypes to maximize available memory

df = df.astype({
    'Year': 'int16',
    'Quarter': 'int16',
    'Month': 'int16',
    'Day_of_Month': 'int16',
    'Day_of_Week': 'int16',
    'DOT_ID_Marketing_Airline': 'int16',
    'Flight_Number_Marketing_Airline': 'int16',
    'DOT_ID_Operating_Airline': 'int16',
    'Flight_Number_Operating_Airline': 'int16',
    'Origin_Airport_ID': 'int16',
    'Origin_Airport_Seq_ID': 'int16',
    'Origin_City_Market_ID': 'int16',
    'Origin_State_Fips': 'int8',
    'Dest_Airport_ID': 'int16',
    'Dest_Airport_Seq_ID': 'int16',
    'Dest_City_Market_ID': 'int16',
    'Dest_State_Fips': 'int16',
    'CRS_Dep_Time': 'int16'
})

In [29]:
df.dtypes

Year                                 int16
Quarter                              int16
Month                                int16
Day_of_Month                         int16
Day_of_Week                          int16
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int16
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int16
Operating_Airline                   object
DOT_ID_Operating_Airline             int16
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int16
Origin_Airport_ID                    int16
Origin_Airport_Seq_ID                int16
Origin_City_Market_ID                int16
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                     int8
Origin_State_Name                   object
Dest_Airpor

In [30]:
df.shape

(7278739, 65)

In [31]:
df.head()

Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2023        1      1            22            7  2023-01-22   
1  2023        1      1            22            7  2023-01-22   
2  2023        1      1            22            7  2023-01-22   
3  2023        1      1            22            7  2023-01-22   
4  2023        1      1            22            7  2023-01-22   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                             1447   
1                          B6                             1451   
2                          B6                             1453   
3                          B6                             1454   
4                          B6                             1455   

  Operating_Airline   DOT_ID_Operating_Airline IATA_Code_Operating_Airline  \
0                 B6                     20409                          B6   
1                 B6                     20409                          B6   
2                 B6                     20409                          B6   
3                 B6                     20409                          B6   
4                 B6                     20409                          B6   

  Tail_Number  Flight_Number_Operating_Airline  Origin_Airport_ID  \
0      N636JB                             1447              12197   
1      N3112J                             1451              10721   
2      N507JT                             1453              11697   
3      N956JT                             1454              14843   
4      N249JB                             1455              10721   

   Origin_Airport_Seq_ID  Origin_City_Market_ID Origin     Origin_City_Name  \
0                 -25482                  31703    HPN     White Plains, NY   
1                  23526                  30721    BOS           Boston, MA   
2                  -9942                  32467    FLL  Fort Lauderdale, FL   
3                 -23022                 -30717    SJU         San Juan, PR   
4                  23526                  30721    BOS           Boston, MA   

  Origin_State  Origin_State_Fips Origin_State_Name  Dest_Airport_ID  \
0           NY                 36          New York            15304   
1           MA                 25     Massachusetts            13204   
2           FL                 12           Florida            14843   
3           PR                 72       Puerto Rico            11697   
4           MA                 25     Massachusetts            11278   

   Dest_Airport_Seq_ID  Dest_City_Market_ID Dest       Dest_City_Name  \
0                23074               -32341  TPA            Tampa, FL   
1                 9682                31454  MCO          Orlando, FL   
2               -23022               -30717  SJU         San Juan, PR   
3                -9942                32467  FLL  Fort Lauderdale, FL   
4                13693                30852  DCA       Washington, DC   

  Dest_State  Dest_State_Fips Dest_State_Name  CRS_Dep_Time  Dep_Time  \
0         FL               12         Florida           800     756.0   
1         FL               12         Florida          1955    2048.0   
2         PR               72     Puerto Rico          1727    1734.0   
3         FL               12         Florida          1140    1147.0   
4         VA               51        Virginia           700     652.0   

   Dep_Delay  Dep_Delay_Minutes  Dep_Del_15  Departure_Delay_Groups  \
0       -4.0                0.0         0.0                    -1.0   
1       53.0               53.0         1.0           

In [32]:
# Save Clean 2023 Data

df.to_csv(os.path.join(path, 'clean_flights2023.csv'))